File: server/src/collectors/ndbOpen.js
/**
 * NDB Open Data collector (ndb-open).
 *
 * MHLW's National Database (NDB) open data
 * (https://www.mhlw.go.jp/stf/seisakunitsuite/bunya/0000177182.html) is
 * published as bulk Excel workbooks per edition — there is no key-free
 * structured API. This collector performs a best-effort reachability probe
 * of the index page and emits a single intel index item. No fabricated
 * claims data is ever emitted; unreachable → `unreachable` tag + live:false.
 *
 * UNVERIFIED: Excel-only dataset, no machine API — probe-only by design.
 */

import { intelEnvelope, intelUid } from '../utils/intelHelpers.js';
import { fetchHead } from './_liveHelpers.js';

const SOURCE_ID = 'ndb-open';
const PROBE_URL = 'https://www.mhlw.go.jp/stf/seisakunitsuite/bunya/0000177182.html';

export default async function collectNdbOpen() {
  const reachable = await fetchHead(PROBE_URL);
  const items = [{
    uid: intelUid(SOURCE_ID, 'portal'),
    title: 'NDB Open Data (health-insurance claims)',
    summary: 'レセプト情報・特定健診等オープンデータ',
    body: 'MHLW National Database open data — aggregated health-insurance claims and specific health checkups, published as Excel workbooks per edition.',
    link: PROBE_URL,
    language: 'ja',
    published_at: new Date().toISOString(),
    tags: ['health', 'statistics', reachable ? 'reachable' : 'unreachable'],
    properties: {
      org: 'MHLW',
      dataset: 'NDB Open Data',
      format: 'xlsx',
      reachable,
    },
  }];

  return intelEnvelope({
    sourceId: SOURCE_ID,
    items,
    live: false,
    extraMeta: { source: reachable ? 'ndb_open_probe' : 'ndb_open_unavailable', probe: true, reachable },
    description: 'MHLW NDB open data index reachability (probe-only)',
  });
}

File: native/collectors/health/sources/ndb_open.c
/* collectors/health/sources/ndb_open.c
 * Port of server/src/collectors/ndbOpen.js — fetchHead reachability probe →
 * ONE intel item (uid = ndb-open|portal == intelUid(SOURCE_ID,'portal')).
 * NDB open data is bulk Excel only; no machine API. ZERO data rows correct. */
#include "../../../source.h"
#include "../../../lib/probe.h"
#include "../../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>

#define PROBE_URL "https://www.mhlw.go.jp/stf/seisakunitsuite/bunya/0000177182.html"

static int run(const source_ctx *ctx, intel_sink *sink) {
  int live = probe_head(ctx->http, PROBE_URL);
  char now[32]; probe_iso_now(now, sizeof now);

  /* tags: ['health','statistics', live?'reachable':'unreachable'] */
  cJSON *tags = cJSON_CreateArray();
  cJSON_AddItemToArray(tags, cJSON_CreateString("health"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("statistics"));
  cJSON_AddItemToArray(tags, cJSON_CreateString(live ? "reachable" : "unreachable"));
  char *tj = cJSON_PrintUnformatted(tags);

  /* properties: { org, dataset, format, reachable } — EXACT JS key order */
  cJSON *p = cJSON_CreateObject();
  cJSON_AddStringToObject(p, "org", "MHLW");
  cJSON_AddStringToObject(p, "dataset", "NDB Open Data");
  cJSON_AddStringToObject(p, "format", "xlsx");
  cJSON_AddBoolToObject(p, "reachable", live);
  char *pj = cJSON_PrintUnformatted(p);

  intel_item it = {0};
  it.remote_key   = "portal";                 /* → uid ndb-open|portal */
  it.title        = "NDB Open Data (health-insurance claims)";
  it.summary      = "レセプト情報・特定健診等オープンデータ";
  it.body         = "MHLW National Database open data — aggregated health-insurance "
                    "claims and specific health checkups, published as Excel workbooks per edition.";
  it.link         = PROBE_URL;
  it.lang         = "ja";
  it.published_at = now;
  it.record_type  = "ndb-open";
  it.tags_json    = tj;
  it.properties_json = pj;
  int rc = sink->emit(sink, &it);

  free(tj); free(pj);
  cJSON_Delete(tags); cJSON_Delete(p);
  fprintf(stderr, "[ndb-open] probe reachable=%d\n", live);
  return rc >= 0 ? 0 : -1;
}

static const source_def ndb_open_def = {
  .id = "ndb-open", .collector = "health",
  .name = "NDB Open Data", .name_ja = "NDB オープンデータ",
   .update_interval_sec = 604800, .run = run };
REGISTER_SOURCE(ndb_open_def)
